server/reports/utils.js

```diff
@@ -24,7 +24,35 @@ const cleanUpObject = (node) => {
   return newNode;
 };
 
+const getNestedValue = (node, key) => {
+  key = key + "";
+  return key.split(".").reduce((value, part) => value && value[part], node);
+};
+
+const sortNodes = (nodes, sort, dir) => {
```

Comment:
I would actually prefer to use sortBy method from lodash, also I am not a big fan of drilling through the "." separator string to access nested values. mainly because it's slightly inneficient, adds complexity that can be avoided, hard to validate all use cases, and harder to transition to typescript.

I think ideally it would look like this:

Input to convert_data:

`{ data: { /* nested, ungrouped etc * / } , arguments: { .. } }`

First there is conversion method, let's say 'convert', which will take in above structure, and return the same structure but with another key, 'convertedData' or something like that, which is data ready for presentation with template.html, so acess to the data and sorting would be much simpler:

 `{% for stockLine in data.data.stockLines.nodes %}` -> `{% for row in data.convertedData %}`

`{{stockLine.item.code}}` -> `{{row.itemCode}}`

This way we can also sort on the whole object simply with

`const sorted = sortBy(convertedData, [sort])`
`if (dir === 'DESC') sorted.reverse();`

In terms of helper methods I would go for this abstraction:

``` 
function convert_data() {
  const res = JSON.parse(Host.inputString());
  const unsortedConvertedData = convert(res); // from convert.js 
  const convertedData = sort(unsortedConvertedData, DEFAULT_SORT_KEY, DEFAULT_SORT_DIRECTION) // from default_sort.js
  Host.outputString(JSON.stringify( { ..res, convertedData }));
}
```

Not requesting any of the above changes in this PR, but I do think it need to be done at some stage

Reply:
That is awesome, agree the nested . structure looks fragile.